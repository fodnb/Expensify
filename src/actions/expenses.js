import uuid from "uuid";
import database from "../../src/firebase/firebase";
//components call action generator
//action generator returns object
//component calls action generator dispatches function
// function runs (has the ability to dispatch other and do whatever it wants)


export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
})

// REMOVE EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: "REMOVE_EXPENSE",
    id 
});

export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch)=>{
        const {
            description = "",
            note = '',
            createdAt = 0,
            amount = 0
        } = expenseData;

        const expense = { description, note, createdAt, amount};

       return database.ref('expenses').push(expense)
            .then((snapshot)=>{
                dispatch(addExpense({
                    id: snapshot.key,
                    ...expense
                }))
            });


    }
}