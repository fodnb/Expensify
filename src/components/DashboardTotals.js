// import selectExpensesTotals from '../selectors/expenses-total.js';


// const expenses = [{
//     id: "1",
//     description: "gum",
//     note: '',
//     amount: 195
// },
// {
//     id: "2",
//     description: "rent",
//     note: '',
//     amount: 170000.00
// },
// {
//     id: "3",
//     description: "credit card",
//     note: '',
//     amount: 4000
// }    
// ]

export default(expenses) =>{

const getExpenseTotal = (expenses) =>{
    expenseTotal = expenses.map((expense)=> {
        return totalsArray.push(expense.amount);
    })
    total = totalsArray.reduce((prev, cur)=> {
        return total = prev + cur;
    })
    
    return total
}

let totalsArray = [], total, expenseTotal;

{expenses[0].amount ? getExpenseTotal(expenses) : total = 0}

return total;

}