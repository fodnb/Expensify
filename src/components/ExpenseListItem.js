import React from 'react';
import {Link} from 'react-router-dom';

const ExpenseListItem = ({ id, description, createdAt, amount}) => (
    <div>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>{createdAt}</p>
        <p>{amount}</p>
        
    </div>
);

export default ExpenseListItem;  