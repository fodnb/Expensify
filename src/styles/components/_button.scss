.button {
    background: $blue;
    color: white;
    padding: $s-size;
    border: none;
    margin: 0;
    font: {
        size: $font-size-large;
        weight: 300;
    }
}

.button--link {
    background: none;
}

.secondary {
    background: #888;
}