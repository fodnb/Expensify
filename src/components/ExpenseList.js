import React from "react";
import { connect } from 'react-redux';
import ExpenseListItem  from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (

    <div>
        {props.expenses.length === 0 ? (<p> No Expenses </p>): (
            props.expenses.map((expense)=>(
                <ExpenseListItem key={expense.id} {...expense}/>
            ))   
        )}
        <h1>expense list</h1>
        
    </div>
);

const mapStoreToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStoreToProps)(ExpenseList);
