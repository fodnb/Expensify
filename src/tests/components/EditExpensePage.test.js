import React from "react";
import { shallow } from 'enzyme';
import {EditExpensePage} from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";


let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach (()=> {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = {
        push: jest.fn()
    }
    wrapper = shallow(
                        <EditExpensePage 
                        history={history} 
                        startEditExpense={startEditExpense} 
                        startRemoveExpense={startRemoveExpense}
                        expense={expenses[2]}
                        />)
});


test('should render edit expense page correctly',()=> {

    expect(wrapper).toMatchSnapshot();

});

test('should handle on submit correctly', ()=>{

    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
})

test('should handle on remove correctly', ()=>{
    wrapper.find('button').simulate("click");
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expenses[2].id})
})