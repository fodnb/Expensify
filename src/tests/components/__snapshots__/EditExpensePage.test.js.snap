// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditExpensePage
    editExpense={[MockFunction]}
    expense={
      Object {
        "amount": 4000,
        "createdAt": 345600000,
        "description": "credit card",
        "id": "3",
        "note": "",
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    removeExpense={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ExpenseForm
          expense={
            Object {
              "amount": 4000,
              "createdAt": 345600000,
              "description": "credit card",
              "id": "3",
              "note": "",
            }
          }
          onSubmit={[Function]}
        />,
        <button
          onClick={[Function]}
        >
          REMOVE 
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "expense": Object {
            "amount": 4000,
            "createdAt": 345600000,
            "description": "credit card",
            "id": "3",
            "note": "",
          },
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "REMOVE ",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "REMOVE ",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ExpenseForm
            expense={
              Object {
                "amount": 4000,
                "createdAt": 345600000,
                "description": "credit card",
                "id": "3",
                "note": "",
              }
            }
            onSubmit={[Function]}
          />,
          <button
            onClick={[Function]}
          >
            REMOVE 
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "expense": Object {
              "amount": 4000,
              "createdAt": 345600000,
              "description": "credit card",
              "id": "3",
              "note": "",
            },
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "REMOVE ",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "REMOVE ",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
